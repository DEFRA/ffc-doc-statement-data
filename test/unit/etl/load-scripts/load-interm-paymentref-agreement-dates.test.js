const db = require('../../../../app/data')
const { loadIntermPaymentrefAgreementDates } = require('../../../../app/etl/load-scripts/load-interm-paymentref-agreement-dates')

jest.mock('../../../../app/data', () => ({
  sequelize: {
    query: jest.fn()
  }
}))

describe('loadIntermPaymentrefAgreementDates', () => {
  const startDate = '2023-01-01'
  const transaction = {}

  beforeEach(() => {
    db.sequelize.query.mockClear()
  })

  test('should call sequelize.query with correct SQL and parameters', async () => {
    await loadIntermPaymentrefAgreementDates(startDate, transaction)

    expect(db.sequelize.query).toMatchSnapshot()
  })

  test('should handle errors thrown by sequelize.query', async () => {
    db.sequelize.query.mockRejectedValue(new Error('Query failed'))

    await expect(loadIntermPaymentrefAgreementDates(startDate, transaction)).rejects.toThrow('Query failed')
  })
})
