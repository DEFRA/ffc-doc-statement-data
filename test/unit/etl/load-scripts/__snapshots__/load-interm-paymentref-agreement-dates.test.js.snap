// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadIntermPaymentrefAgreementDates should call sequelize.query with correct SQL and parameters 1`] = `
[MockFunction] {
  "calls": [
    [
      "
    INSERT INTO etlIntermPaymentrefAgreementDates (
      paymentRef, agreementStart, agreementEnd
    )
    SELECT
      DA.paymentRef,
      (SELECT agreementStart FROM etlIntermApplicationContract IAC WHERE IAC.contractId = CA.contractId LIMIT 1),
      (SELECT agreementEnd FROM etlIntermApplicationContract IAC WHERE IAC.contractId = CA.contractId LIMIT 1)
    FROM etlIntermFinanceDax DA
    INNER JOIN etlStageCssContractApplications CA ON CA.applicationId = DA.claimId
    INNER JOIN etlIntermApplicationContract IAC ON IAC.contractId = CA.contractId
    WHERE IAC.agreementStart IS NOT NULL
      AND IAC.agreementEnd IS NOT NULL
      AND (
        DA.etlInsertedDt > :startDate
        OR CA.etlInsertedDt > :startDate
        OR IAC.etlInsertedDt > :startDate
      )
    GROUP BY DA.paymentRef, CA.contractId
    ON CONFLICT (paymentRef, agreementStart, agreementEnd) DO NOTHING;
  ",
      {
        "raw": true,
        "replacements": {
          "startDate": "2023-01-01",
        },
        "transaction": {},
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
